### PR TITLE
Refactor beam search to use gluon hybrid blocks

### PR DESCRIPTION
**Pull request by fhieber**
This refactors beam search to group fixed-size operations in beam search into cached ops through Gluon HybridBlocks.
My testing showed ~3% speed improvement. Not much, but consistent.

Another change that is included here is to not use columns from `sequences` to pass into the decoder step module, but use `best_word_indices` from the previous iteration. NDArray indexing seems expensive and ideally we should aim for avoiding all indexing ops in an iteration.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by mjpost** [comment]
Can you pull in the bug fix I just pushed and make sure the tests still pass? Some of the tests were broken for constrained decoding, which this might affect.

**Comment by fhieber** [comment]
Yes will do that soon

**Comment by fhieber** [comment]
Tests still pass. Also ran the system tests but not sure if they even cover constraints.

File: sockeye/inference.py [hunk 1]
```diff
@@ -854,12 +854,21 @@ class LengthPenalty:
...
-    def __call__(self, lengths: Union[mx.nd.NDArray, int, float]) -> Union[mx.nd.NDArray, float]:
+    def hybrid_forward(self, F, lengths):
+        if self.alpha == 0.0:
+            return F.ones_like(lengths)
+        else:
+            numerator = self.beta + lengths if self.beta != 0.0 else lengths
+            numerator = numerator ** self.alpha if self.alpha != 1.0 else numerator
+            return numerator / self.denominator
+
+    def get(self, lengths: Union[mx.nd.NDArray, int, float]) -> Union[mx.nd.NDArray, float]:
```
Comment (tdomhan) [inline sockeye/inference.py]: Could we merge `get` and `hybrid_forward` somehow?  e.g.: 
```
   def get(...):
       return hybird_forward(None, lengths)
   def hybrid_forward(F, lengths):
        if F is None:
             return 1
         else:
             return F.ones_like(...)
```

File: sockeye/inference.py [hunk 2]
```diff
@@ -1339,10 +1362,12 @@ def _beam_search(self,
         # General data structure: each row has batch_size * beam blocks for the 1st sentence, with a full beam,
         # then the next block for the 2nd sentence and so on
 
+        best_word_indices = mx.nd.full((self.batch_size * self.beam_size,), val=self.start_id, ctx=self.context,
+                                       dtype='int32')
         # sequences: (batch_size * beam_size, output_length), pre-filled with <s> symbols on index 0
         sequences = mx.nd.full((self.batch_size * self.beam_size, max_output_length), val=C.PAD_ID, ctx=self.context,
                                dtype='int32')
-        sequences[:, 0] = self.start_id
+        sequences[:, 0] = best_word_indices
```
Comment (tdomhan) [inline sockeye/inference.py]: is this faster, or why did you change this?

File: sockeye/inference.py [hunk 3]
```diff
@@ -1464,35 +1484,38 @@ def _beam_search(self,
...
-            finished = all_finished
+            finished, scores_accumulated = self._normalize_finished.forward(best_word_indices,
+                                                                            finished,
+                                                                            scores_accumulated,
+                                                                            lengths)
 
             # (6) Prune out low-probability hypotheses. Pruning works by setting entries `inactive`.
             if self.beam_prune > 0.0:
-                inactive = self.prune(scores_accumulated, finished)
+                inactive = self._prune_hyps.forward(scores_accumulated, finished, self.inf_array)
```
Comment (tdomhan) [inline sockeye/inference.py]: why not also pull in the where's below?

File: sockeye/inference.py (hunk 1 above)
Comment (fhieber) [reply sockeye/inference.py]: sure, done

File: sockeye/inference.py (hunk 2 above)
Comment (fhieber) [reply sockeye/inference.py]: yes, as said in the PR text it seems that indexing is slow. This actually gave a small speedup.

File: sockeye/inference.py (hunk 3 above)
Comment (fhieber) [reply sockeye/inference.py]: We just had changed this [recently](https://github.com/awslabs/sockeye/commit/fc440df9cccce7d933b520f320a546cecf82325b) to make the interaction between `inactive` and affected data structures more clear.
I could change it but the trouble is that hybrid where's are more strict w.r.t to shape: thinks like `where((5,), (5,1), (5,)` don't work in HybridBlocks but imperatively apparently. 
I tried for a while to get the shapes of all the auxiliary arrays and `scores_accumulated` etc. to work fine, but its adding more reshapes here and there.

File: sockeye/inference.py (hunk 2 above)
Comment (fhieber) [reply sockeye/inference.py]: actually, scratch that. I misread the position you were commenting on.
I can revert this.

File: sockeye/inference.py (hunk 3 above)
Comment (fhieber) [reply sockeye/inference.py]: will do

File: sockeye/inference.py (hunk 2 above)
Comment (tdomhan) [reply sockeye/inference.py]: I mean it's fine, I was just wondering.
